build-system/tasks/check-types.js

```diff
@@ -23,87 +25,150 @@ const {
 } = require('../compile/debug-compilation-lifecycle');
 const {cleanupBuildDir, closureCompile} = require('../compile/compile');
 const {compileCss} = require('./css');
+const {cyan, green, yellow, red} = require('kleur/colors');
 const {extensions, maybeInitializeExtensions} = require('./extension-helpers');
 const {log} = require('../common/logging');
 const {typecheckNewServer} = require('../server/typescript-compile');
 
+const EXTERNS_GLOB = 'src/core{,/**}/*.extern.js';
+
+/**
+ * Generates a list of source file paths for extensions to type-check
+ * Must be run after `maybeInitializeExtensions`
+ * @function
+ * @return {!Array<string>}
+ */
+const getExtensionSrcPaths = () =>
+  Object.values(extensions)
+    .filter((ext) => !ext.noTypeCheck)
+    .map(({name, version}) => `extensions/${name}/${version}/${name}.js`)
+    .sort();
+
+/**
+ * The main configuration location to add/edit targets for type checking.
+ * Properties besides `entryPoints` are passed on to `closureCompile` as
+ * options. * Values may be objects or functions, as some require initialization
+ * or filesystem access and shouldn't be run until needed.
+ * @type {Object<string, Object|function():Object>}
+ */
+const TYPE_CHECK_TARGETS = {
+  'src': {
+    entryPoints: [
```

Comment:
Would it make more sense to use our bundle definitions as our typecheck targets instead of recreating a somewhat similar map? The primary entrypoints are in `bundles.config.js` and the extensions are in `bundles.config.extensions.js`.

In the short-term, I see how that would be difficult for enforcing `src/core`  files, but in the long terms it seems more maintainable (and hopefully less code!).

Reply:
The short answer to this is: I don't know. To some degree, that would seem to make sense. When the whole repo is passing, we can enable the type--checking CC flag by default, and that will just get added into the `dist` process. But that's a bit "messier" to get working, as opposed to locking down type-safety one directory at a time. If we build this up so the whole repo is type-checked successfully, would there be any added value in type-checking the binaries themselves? (I'm guessing yes, but am not certain)

Reply:
Partial typechecking will not be valuable once everything is fixed up. The question is until we get there, what is more valuable? Fixing directory by directory where we have to code each in a one-off way? Or Binary by binary?  I don't have _that_ strong of an opinion since I figured adding in one-off directories is not that much work

Reply:
I get the sense that in practice, locking down on directory at a time (and including the check in CI) will make it easiest to take multiple bites at the apple and get the whole repo compliant, whereas we have to get a whole binary compliant to enable it during CI and for most of the binaries that entails most of `src`